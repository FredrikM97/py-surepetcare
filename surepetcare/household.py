import logging
from surepetcare.command import Command
from surepetcare.const import API_ENDPOINT_PRODUCTION, API_ENDPOINT_V1
from surepetcare.devices import load_device_class
from surepetcare.devices.pet import Pet
from surepetcare.enums import ProductId

logger = logging.getLogger(__name__)

class Household:
    def __init__(self, data: dict):
        self.data = data
        self.id = data["id"]
        # Add other fields as needed

    def get_pets(self):
        def parse(response):
            if response['status'] == 304:
                return self.data.get("pets", [])
            pets = [Pet(p) for p in response["data"]]
            self.data['pets'] = pets
            return pets
            

        return Command(
            method="GET", endpoint=f"{API_ENDPOINT_V1}/pet", params={"HouseholdId": self.id}, callback=parse
        )

    def get_devices(self):
        def parse(response):
            logger.info(f"Parsing devices for household {self.id}: {response}")
            if response['status'] == 304:
                logger.info("Returning cached devices")
                return self.data.get("devices",[])
            devices = []
            for device in response["data"]:
                if device["product_id"] in set(ProductId):
                    devices.append(load_device_class(device["product_id"])(device))
            self.data['devices'] = devices
            return devices

        return Command(
            method="GET",
            endpoint=f"{API_ENDPOINT_PRODUCTION}/device",
            params={"HouseholdId": self.id},
            callback=parse,
        )

    @staticmethod
    def get_households():
        def parse(response):
            return [Household(h) for h in response["data"]]

        return Command(method="GET", endpoint=f"{API_ENDPOINT_PRODUCTION}/household", params={}, callback=parse, reuse=False)

    @staticmethod
    def get_household(household_id: int):
        return Command(method="GET", endpoint=f"{API_ENDPOINT_PRODUCTION}/household/{household_id}", reuse=False)

    @staticmethod
    def get_product(product_id: ProductId, device_id: int):
        """TODO: Move to devices instead"""
        return Command(
            method="GET", endpoint=f"{API_ENDPOINT_PRODUCTION}/product/{product_id}/device/{device_id}/control", reuse=False
        )
