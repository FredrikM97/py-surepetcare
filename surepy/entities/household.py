from surepy.const import API_ENDPOINT_V1
from surepy.const import API_ENDPOINT_V2
from surepy.devices import load_device_class
from surepy.devices.base import SurepyDevice
from surepy.entities.pet import Pet
from surepy.enums import ProductId
from surepy.helper import AbstractHasGet


class HouseholdMixin(AbstractHasGet):
    async def get_household(self, household_id: int):
        return await self.get(f"{API_ENDPOINT_V1}/household/{household_id}")

    async def get_households(self):
        return (await self.get(f"{API_ENDPOINT_V1}/household"))["data"]

    async def get_devices(self, household_ids: list[int]) -> list[SurepyDevice]:
        """Get devices for a list of household IDs."""
        products = set(ProductId)

        devices = []
        for household_id in household_ids:
            household_devices = (
                await self.get(f"{API_ENDPOINT_V1}/device", params={"HouseholdId": household_id})
            )["data"]
            for device in household_devices:
                if device["product_id"] in products:
                    devices.append(load_device_class(ProductId(device["product_id"]))(self, device))

        return devices

    async def get_product(self, product_id: ProductId, device_id: int):
        return await self.get(f"{API_ENDPOINT_V2}/product/{product_id}/device/{device_id}/control")

    async def get_pets(self, household_id: int) -> list[Pet]:
        pets = []
        for pet in (await self.get(f"{API_ENDPOINT_V1}/pet", params={"HouseholdId": household_id}))["data"]:
            pets.append(Pet(self, pet))
        return pets
