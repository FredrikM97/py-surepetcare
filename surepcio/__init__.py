import logging

from surepcio.security.redact import RedactSensitiveFilter
from surepcio.client import SurePetcareClient
from surepcio.household import Household

handler = logging.StreamHandler()
handler.addFilter(RedactSensitiveFilter())

logging.basicConfig(level=logging.INFO, handlers=[handler])

