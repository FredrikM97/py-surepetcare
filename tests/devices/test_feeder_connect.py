import pytest
from syrupy.assertion import SnapshotAssertion

from surepcio.household import Household
from tests.mock_helpers import MockClient
from tests.mock_helpers import recursive_dump


@pytest.fixture
def device_file():
    return "tests/fixture/feeder_connect.json"


@pytest.fixture
def household_file():
    return "tests/fixture/household.json"


@pytest.mark.asyncio
async def test_snapshot(snapshot: SnapshotAssertion, household_file, device_file):
    snapshot.snapshot_dir = "tests/snapshots"
    client = MockClient(fixture_file=device_file)
    client.set_mock_response(household_file)
    household = await client.api(Household.get_household(7777))
    client.reset()
    device = (await client.api(household.get_devices()))[0]
    await client.api(device.refresh())
    assert recursive_dump(device) == snapshot
