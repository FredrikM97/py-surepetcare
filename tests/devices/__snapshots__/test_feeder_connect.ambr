# serializer version: 1
# name: test_snapshot[device_names0]
  dict({
    'control': dict({
      'bowls': dict({
        'settings': list([
          dict({
            'food_type': <FoodType.WET: 1>,
            'target': 0,
          }),
          dict({
            'food_type': <FoodType.DRY: 2>,
            'target': 0,
          }),
        ]),
        'type': <BowlType.TWO_SMALL: 4>,
      }),
      'fast_polling': False,
      'lid': dict({
        'close_delay': <CloseDelay.FASTER: 0>,
      }),
      'tags': list([
        dict({
          'created_at': datetime.datetime(2023, 10, 1, 14, 34, 49, tzinfo=TzInfo(UTC)),
          'device_id': 269654,
          'id': 60978,
          'index': 1,
          'profile': 2,
          'updated_at': datetime.datetime(2023, 10, 1, 14, 34, 49, tzinfo=TzInfo(UTC)),
          'version': 0,
        }),
      ]),
      'tare': <Tare.TWO_SMALL: 2>,
      'training_mode': <FeederTrainingMode.DISABLED: 0>,
    }),
    'entity_info': dict({
      'household_id': 7777,
      'id': 269654,
      'name': 'Maui matlåda (old) ',
      'parent_device_id': None,
      'photo': None,
      'product_id': 4,
      'tag': None,
      'tag_id': None,
    }),
    'status': dict({
      'battery': 5.046,
      'bowl_status': list([
        dict({
          'current_weight': 32.16,
          'fill_percent': 100,
          'food_type': <FoodType.WET: 1>,
          'last_fill_weight': 32.16,
          'last_filled_at': datetime.datetime(2025, 7, 7, 10, 29, 29, tzinfo=TzInfo(UTC)),
          'last_zeroed_at': datetime.datetime(2025, 5, 23, 10, 25, 29, tzinfo=TzInfo(UTC)),
          'position': <BowlPosition.ONE: 0>,
          'substance_type': <SubstanceType.FOOD: 2>,
        }),
        dict({
          'current_weight': -27.2,
          'fill_percent': None,
          'food_type': <FoodType.WET: 1>,
          'last_fill_weight': -27.2,
          'last_filled_at': datetime.datetime(2025, 7, 7, 10, 29, 29, tzinfo=TzInfo(UTC)),
          'last_zeroed_at': datetime.datetime(2025, 5, 23, 10, 25, 29, tzinfo=TzInfo(UTC)),
          'position': <BowlPosition.TWO: 1>,
          'substance_type': <SubstanceType.FOOD: 2>,
        }),
      ]),
      'learn_mode': False,
      'online': True,
      'signal': dict({
        'device_rssi': -79,
      }),
      'version': dict({
        'device': dict({
          'firmware': '406.437',
          'hardware': '9',
        }),
      }),
    }),
    'timezone': 'Europe/Stockholm',
  })
# ---
# name: test_snapshot_set_bowls_command[device_names0]
  dict({
    'control': dict({
      'bowls': dict({
        'settings': list([
          dict({
            'food_type': <FoodType.WET: 1>,
            'target': 50,
          }),
          dict({
            'food_type': <FoodType.WET: 1>,
            'target': 100,
          }),
        ]),
        'type': <BowlType.TWO_SMALL: 4>,
      }),
      'fast_polling': False,
      'lid': dict({
        'close_delay': <CloseDelay.SLOWER: 20>,
      }),
      'tags': None,
      'tare': None,
      'training_mode': <FeederTrainingMode.STEP_1: 1>,
    }),
    'entity_info': dict({
      'household_id': 7777,
      'id': 269654,
      'name': 'Maui matlåda (old) ',
      'parent_device_id': None,
      'photo': None,
      'product_id': 4,
      'tag': None,
      'tag_id': None,
    }),
    'status': dict({
      'battery': None,
      'bowl_status': None,
      'learn_mode': None,
      'online': None,
      'signal': None,
      'version': None,
    }),
    'timezone': 'Europe/Stockholm',
  })
# ---
